src/gui/qgsmaptoolcapture.cpp

```diff
@@ -811,9 +811,8 @@ QgsPoint QgsMapToolCapture::mapPoint( const QgsMapMouseEvent &e ) const
     if ( e.isSnapped() )
     {
       const QgsPointLocator::Match match = e.mapPointMatch();
-      const QgsWkbTypes::Type snappedType = match.layer()->wkbType();
 
-      if ( QgsWkbTypes::hasZ( snappedType ) )
+      if ( match.layer() && QgsWkbTypes::hasZ( match.layer()->wkbType() ) )
       {
         const QgsFeature ft = match.layer()->getFeature( match.featureId() );
```

Comment:
Isn't that a very expensive operation after we have already fetched the geometry with Z in the above code and could cache it?
I realize that the code already exists like that, just wanting to see if there's potentially a possibility to improve (e.g. `match.getPoint()` and `match.getPointXY()`, where the first one will only request the data provider if it's not cached.)

Reply:
I only fixed a potential crash on this part. There maybe optimizations to be made, but all cases must be studied with the Z and this is out of the scope of this PR.

Reply:
yep, just exploring ideas here :)